src/Microsoft.AspNetCore.Http/PipeWriterAdapter.cs

```diff
@@ -0,0 +1,287 @@
+        public override void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            // Note this currently doesn't do anything.
+            // Implementing completions/callbacks would require creating a PipeReaderAdapter.
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            return FlushAsyncInternal(cancellationToken);
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            if (_bytesWritten == 0)
+            {
+                return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+            }
+
+            // Wrap the provided cancellationToken with an internal one
+            // to allow CancelPendingFlush to cancel writes and flushes
+            if (_currentFlushCanceled)
+            {
+                // If CancelPendingFlush was already called, we need to return a FlushResult with
+                // Canceled. At this point, we create a fresh CTS that isn't canceled.
+                var result = new FlushResult(isCanceled: true, IsCompletedOrThrow());
+                InternalTokenSource = new CancellationTokenSource();
+                _currentFlushCanceled = false;
+                return result;
+            }
+
+            // As an optimization, we only create a new CTS if the previous one wasn't canceled.
+            // Otherwise, we would need to create a new CTS for every call to FlushAsync.
+            var joinedToken = CancellationTokenSource.CreateLinkedTokenSource(
+                cancellationToken,
+                InternalTokenSource.Token).Token;
+
+            try
+            {
+                // Write all completed segments and whatever remains in the current segment
+                // and flush the result.
+                if (_completedSegments != null)
+                {
+                    var count = _completedSegments.Count;
+                    for (var i = 0; i < count; i++)
+                    {
+                        var segment = _completedSegments[i];
+                        await _writingStream.WriteAsync(segment.Buffer, 0, segment.Length, joinedToken);
+                    }
+                }
+
+                if (_currentSegment != null)
+                {
+                    await _writingStream.WriteAsync(_currentSegment, 0, _position, joinedToken);
+                }
+
+                await _writingStream.FlushAsync(joinedToken);
+            }
+            catch (OperationCanceledException)
+            {
+                // Catch any cancellation and translate it into setting isCanceled = true
+                return new FlushResult(isCanceled: true, IsCompletedOrThrow());
```

Comment:
If `CancellationToken cancellationToken` was cancelled we need to throw the exception.